cmd/thor/node/backer_loop.go

```diff
@@ -0,0 +1,427 @@
+			break
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C():
+			continue
+		}
+	}
+
+	var scope event.SubscriptionScope
+	defer scope.Close()
+
+	newProposalCh := make(chan *comm.NewProposalEvent)
+	scope.Track(n.comm.SubscribeProposal(newProposalCh))
+
+	newAcceptedCh := make(chan *comm.NewAcceptedEvent)
+	scope.Track(n.comm.SubscribeAccepted(newAcceptedCh))
+
+	unknownTicker := time.NewTicker(time.Duration(1) * time.Second)
+	defer unknownTicker.Stop()
+
+	var st *status
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case ev := <-newProposalCh:
+			p := ev.Proposal
+			parent := n.repo.BestBlock()
+
+			//  only accept proposal that are within 3 rounds.
+			if math.Abs(float64(p.Number())-float64(parent.Header().Number()+1)) > 3 {
+				continue
+			}
+
+			b, _ := rlp.EncodeToBytes(p)
+			hash := thor.Blake2b(b)
+
+			// skip if proposal already seen(prevent DoS)
+			if seenProposal.Contains(hash) {
+				continue
+			}
+			seenProposal.Add(hash, struct{}{})
+
+			if parent.Header().ID() != p.ParentID {
+				unknownProposal.Set(hash, ev.Proposal)
+				continue
+			}
+
+			if st == nil || st.ParentID() != parent.Header().ID() {
+				new, err := newStatus(n, parent)
+				if err != nil {
+					log.Debug("failed to initiate status", "err", err)
+					continue
+				}
+				st = new
+			}
+
+			if err := n.validateProposer(p, st); err != nil {
+				log.Debug("signer not valid", "err", err)
+				continue
+			}
+
+			if err := n.validateProposal(p, st); err != nil {
+				log.Debug("block Proposal is not valid", "err", err)
+				continue
+			}
+
+			n.comm.BroadcastProposal(p)
+			knownProposal.Set(hash, p, float64(p.Timestamp))
+
+			if st.GetAuthority(n.master.Address()) != nil {
+				signature, err := n.tryBacking(p, st)
+				if err != nil {
+					log.Debug("failed to back a proposal", "err", err)
+					continue
+				}
+
+				accepted := proto.Accepted{
+					ProposalHash: hash,
+					Signature:    signature,
+				}
+
+				seenAccepted.Add(accepted.Hash(), struct{}{})
+				n.comm.BroadcastAccepted(&accepted)
+			}
+		case ev := <-newAcceptedCh:
+			parent := n.repo.BestBlock()
+			// skip if backer signature already seen(prevent DOS)
+			if seenAccepted.Contains(ev.Hash()) {
+				continue
+			}
+			seenAccepted.Add(ev.Hash(), struct{}{})
+
+			if val, _, ok := knownProposal.Get(ev.ProposalHash); ok == true {
+				p := val.(*block.Proposal)
+
+				if parent.Header().ID() != p.ParentID {
+					continue
+				}
+
+				if st == nil || st.ParentID() != parent.Header().ID() {
+					new, err := newStatus(n, parent)
+					if err != nil {
+						log.Debug("failed to initiate status", "err", err)
+						continue
+					}
+					st = new
+				}
+
+				if err := n.validateBackerSignature(ev.Signature, p, st); err != nil {
+					log.Debug("failed to validate backer signature", "err", err)
+					continue
+				}
+
+				n.comm.BroadcastAccepted(ev.Accepted)
+			} else {
+				unknownAccepted.Set(ev.Hash(), ev.Accepted)
```

Comment:
should be validated here?

Reply:
The backer signature can not be validated without the proposal